Source/Diagnostics/Diagnostics.cpp

```diff
@@ -70,51 +79,64 @@ Diagnostics::InitData ()
             } else if ( varnames[comp] == "rho" ){
                 // rho_new is stored in component 1 of rho_fp when using PSATD
 #ifdef WARPX_USE_PSATD
                 MultiFab* rho_new = new MultiFab(*warpx.get_pointer_rho_fp(lev), amrex::make_alias, 1, 1);
-                all_field_functors[lev][comp] = new CellCenterFunctor(rho_new, lev);
+                all_field_functors[lev][comp] = new CellCenterFunctor(rho_new, lev, m_crse_ratio);
 #else
-                all_field_functors[lev][comp] = new CellCenterFunctor(warpx.get_pointer_rho_fp(lev), lev);
+                all_field_functors[lev][comp] = new CellCenterFunctor(warpx.get_pointer_rho_fp(lev), lev, m_crse_ratio);
 #endif
             } else if ( varnames[comp] == "F" ){
-                all_field_functors[lev][comp] = new CellCenterFunctor(warpx.get_pointer_F_fp(lev), lev);
+                all_field_functors[lev][comp] = new CellCenterFunctor(warpx.get_pointer_F_fp(lev), lev, m_crse_ratio);
             } else if ( varnames[comp] == "part_per_cell" ){
-                all_field_functors[lev][comp] = new PartPerCellFunctor(nullptr, lev);
+                all_field_functors[lev][comp] = new PartPerCellFunctor(nullptr, lev, m_crse_ratio);
             } else if ( varnames[comp] == "part_per_grid" ){
-                all_field_functors[lev][comp] = new PartPerGridFunctor(nullptr, lev);
+                all_field_functors[lev][comp] = new PartPerGridFunctor(nullptr, lev, m_crse_ratio);
             } else if ( varnames[comp] == "divB" ){
-                all_field_functors[lev][comp] = new DivBFunctor(warpx.get_array_Bfield_aux(lev), lev);
+                all_field_functors[lev][comp] = new DivBFunctor(warpx.get_array_Bfield_aux(lev), lev, m_crse_ratio);
             } else if ( varnames[comp] == "divE" ){
-                all_field_functors[lev][comp] = new DivEFunctor(warpx.get_array_Efield_aux(lev), lev);
+                all_field_functors[lev][comp] = new DivEFunctor(warpx.get_array_Efield_aux(lev), lev, m_crse_ratio);
             }
         }
 
         AddRZModesToDiags( lev );
 
         // At this point, varnames.size() >= all_field_functors[0].size()
 
+        BoxArray ba = warpx.boxArray(lev);
+        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
+            m_crse_ratio.min() > 0,
+            "Coarsening ratio must be non-zero."
+        );
+        AMREX_ALWAYS_ASSERT_WITH_MESSAGE (
+            ba.coarsenable(m_crse_ratio),
+            "Invalid coarsening ratio for diagnostics. Must be a power of two and an integer divisor of the blocking factor."
+        );
+        // The boxArray is coarsened based on the user-defined coarsening ratio
+        // Note : ba.coarsen assumes lo and hi for src_mf and dst_mf are equal
+        ba.coarsen(m_crse_ratio);
+        //  This assert should only be applied if src and dst mf have same lo and hi
+        AMREX_ALWAYS_ASSERT(ba.size() == warpx.boxArray(lev).size());
```

Comment:
I think this is an incomplete, hence dangerous check. As `ba` is just the result of `coarsen` applied on `warpx.boxArray(lev)`, I think we could simply check in AMReX if boxes in `ba` and `warpx.boxArray(lev)` cover the same physical domain. If so no need to check further, unless AMReX has a function that does a complete check on these BA. 
```suggestion
```